File: fn/src/main/java/org/sqlbuilder/fn/joins/SubCorpus_Sentence.java
package org.sqlbuilder.fn.joins;

import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.fn.RequiresIdFrom;
import org.sqlbuilder.fn.objects.Sentence;
import org.sqlbuilder.fn.objects.SubCorpus;

import java.util.Comparator;
import java.util.HashSet;
import java.util.Set;

public class SubCorpus_Sentence extends Pair<SubCorpus, Integer> implements Insertable
{
	public static final Comparator<SubCorpus_Sentence> COMPARATOR = Comparator.comparing(SubCorpus_Sentence::getFirst, SubCorpus.COMPARATOR).thenComparing(SubCorpus_Sentence::getSecond);

	public static final Set<SubCorpus_Sentence> SET = new HashSet<>();

	// C O N S T R U C T O R

	public static SubCorpus_Sentence make(final SubCorpus subcorpus, final Sentence sentence)
	{
		var ss = new SubCorpus_Sentence(subcorpus, sentence.getID());
		SET.add(ss);
		return ss;
	}

	private SubCorpus_Sentence(final SubCorpus subcorpus, final Integer sentenceid)
	{
		super(subcorpus, sentenceid);
	}

	// I N S E R T

	@RequiresIdFrom(type = SubCorpus.class)
	@Override
	public String dataRow()
	{
		return String.format("%s,%d", first.getSqlId(), second);
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("[SUBCORPUS-SENT subcorpusid=%s sentenceid=%s]", first, second);
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/joins/SubCorpus_Sentence.kt
package org.sqlbuilder.fn.joins

import org.sqlbuilder.annotations.RequiresIdFrom
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.fn.objects.Sentence
import org.sqlbuilder.fn.objects.SubCorpus
import java.util.*

data class SubCorpus_Sentence(
    var subcorpus: SubCorpus,
    var sentenceid: Int,
) : Insertable {

    // I N S E R T

    @RequiresIdFrom(type = SubCorpus::class)
    override fun dataRow(): String {
        return "${subcorpus.getSqlId()},$sentenceid"
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as SubCorpus_Sentence
        return subcorpus == that.subcorpus && sentenceid == that.sentenceid
    }

    override fun hashCode(): Int {
        return Objects.hash(subcorpus, sentenceid)
    }

    // T O S T R I N G

    override fun toString(): String {
        return "[SUBCORPUS-SENT subcorpusid=$subcorpus sentenceid=$sentenceid]"
    }

    companion object {

        @JvmField
        val COMPARATOR: Comparator<SubCorpus_Sentence> = Comparator
            .comparing<SubCorpus_Sentence, SubCorpus>({ it.subcorpus }, SubCorpus.COMPARATOR)
            .thenComparing<Int> { it.sentenceid }

        @JvmField
        val SET = HashSet<SubCorpus_Sentence>()

        fun make(subcorpus: SubCorpus, sentence: Sentence): SubCorpus_Sentence {
            val ss = SubCorpus_Sentence(subcorpus, sentence.iD)
            SET.add(ss)
            return ss
        }
    }
}
